server/datastore/mysql/hosts.go

```diff
@@ -466,6 +467,21 @@ func (ds *Datastore) ListHosts(ctx context.Context, filter fleet.TeamFilter, opt
 }
 
 func (ds *Datastore) applyHostFilters(opt fleet.HostListOptions, sql string, filter fleet.TeamFilter, params []interface{}) (string, []interface{}) {
+	deviceMappingJoin := `LEFT JOIN (
+		SELECT
+			he.host_id,
+			GROUP_CONCAT("[", he.dmap, "]") AS device_mapping
+		FROM (
+			SELECT
+				host_id,
+				GROUP_CONCAT(JSON_OBJECT("email", email, "source", source)) AS dmap
+			FROM
+				host_emails
+			GROUP BY
+				host_id) he
+		GROUP BY
+			host_id) dm ON (h.id = dm.host_id)`
```

Comment:
There seems to be an unnecessary subquery going on here. I think this can be simplified to

```suggestion
	deviceMappingJoin := `LEFT JOIN (
		SELECT
			host_id,
			GROUP_CONCAT('[', JSON_OBJECT('email', email, 'source', source), ']') AS device_mapping
		FROM
			host_emails
		GROUP BY
			host_id) dm ON (h.id = dm.host_id)`
```

Reply:
@michalnicp, thank you for the review! I tried to do it without the additional subquery at first but wasn't able to find a good way to handle the case where there is more than one email per host. For example, your revised query above results in two separate arrays inside the string: 
`[{"email": "user1@example.com", "source": "google_chrome_profiles"}],[{"email": "user2@example.com", "source": "google_chrome_profiles"}]`

The only way I was able to get it to work (without using `JSON_ARRAYAGG`) was to add the second query. My hackaround doesn't feel very elegant to me so definitely open to alternative approaches.

Reply:
Sorry, I copy pasted the wrong snippet. I meant to put

```
	deviceMappingJoin := `LEFT JOIN (
		SELECT
			host_id,
			CONCAT('[', GROUP_CONCAT(JSON_OBJECT('email', email, 'source', source)), ']') AS device_mapping
		FROM
			host_emails
		GROUP BY
			host_id) dm ON dm.host_id = h.id `
```

Notice the use of `CONCAT` with `GROUP_CONCAT`.

Reply:
Oh nice, that does the trick!